### ISSUE TITLE
Define terraform structure

### ISSUE THREAD

**Issue by torbencarstens** [BlindfoldedSurgery/architecture#3]

**Comment by torbencarstens** [comment]
Comment on my vote above:

This results in a somewhat more complicate github actions configuration, I think we could away with defining the organizations (backend config, secrets, ....) in the `inputs` section in the workflow and subsequently reuse this in the `matrix` section of each job.

`strategy.fail-fast: false` and `max-parallel: 1` should be set (`max-parallel: 1` may not need to be set depending on where terraform actually locks the backend, if it's based on the `backend.key` then this doesn't need to be set).

Problem with this approach:

`continue-on-error: false` -> one wrong organization configuration prevents all applies

`continue-on-error: true` -> we'll get #planFailures failures in the next job (apply) -> this is no good. I'm not sure that this can be prevented in a somewhat clean way. We could probably mix in an `if` by setting some failure outputs in the plan step, this is too much of a shaky workaround though IMO.

Based on the [docs the matrix context](https://docs.github.com/en/actions/learn-github-actions/contexts#steps-context) doesn't seem to have a sepcial handling for `result` so GHA doesn't seem to support this use case natively (should be tested though)

**Comment by torbencarstens** [comment]
another thing is that we can't check in `.terraform.lock.hcl` since this is backend specific and we need different backends for each organization since the `backend.key` needs to be set.

**Comment by BjoernPetersen** [comment]
I don't quite understand your point about the terraform lockfile. Are you planning to have the exact same terraform manifests applied for each organization? That doesn't make a lick of sense to me. Why not just have a directory per organization, in the same repo?

To take a step back: why do you want to manage more than this organization in a repo here? I'm assuming it's because they share some infrastructure (most prominently the k8s cluster), but I  don't think that was mentioned anywhere yet.

**Comment by torbencarstens** [comment]
> Are you planning to have the exact same terraform manifests applied for each organization

That is one option yes, I don't really see the downside of re-using the existing .tf files with the relevant resources since the github configuration is pretty much the same everywhere with a few different naming schemes and different repos. So we only need different values.

> Why not just have a directory per organization, in the same repo?

Do you mean definining a local module and then having different terraform/values per organization using said module?

This could work I guess.

>  why do you want to manage more than this organization in a repo here?

This is simply one option, for me this reduces overhead on managing the integrity of the terraform files.

Lets say we've a `terraform-github` repo for each organization, if we change the structure/fix a bug/do version upgrades/[...] we have to change it in #organizationCount places (currently 4).

---

There are still a lot of different options not mentioned yet in the issue options, we could also use tools like terragrunt or change the implementation detail of an existing option.

**Comment by BjoernPetersen** [comment]
I think I'm okay with quadruplicating the github-terraform repo for now. Encapsulating most of our defaults in a [terraform module](https://github.com/BlindfoldedSurgery/terraform-repo-module) worked out pretty well, so the duplication (within Terraform) isn't too bad IMO. If the current solution remains stable, I'd be open to centralize them all in a single BS repo, but I don't want to commit to an abstraction before we've gained some more experience with GitHub+Terraform.

For Cloudflare, I feel like we might need a mixed approach. Org-specific domains (woog.life, tim[...].consulting, etc.) should be managed in their respective org. Other general-shitposting-use domains should probably all be managed in a single repo to avoid the overhead of managing multiple repos.

**Comment by torbencarstens** [comment]
Yeah, I agree on both fronts. Has been working fairly well so far in the other orgs.
